File: impl/typescript/src/canonical.ts
/** Canonicalization and SHA-256 hashing (SPEC §9). Columnar-invariant (§9.4a). */

import { createHash } from "node:crypto";
import type { Message, PairlRecord } from "./model.js";

const HEADER_ORDER = ["v", "id", "mid", "sid", "ts", "p", "parent", "root", "deps", "budget", "quota"];
const ATOM_OK = /^[A-Za-z0-9:._/@+-]+$/;

function fmtValue(v: string): string {
  return v === "" || !ATOM_OK.test(v) ? `"${v.replace(/"/g, '\\"')}"` : v;
}

export function serializeRecord(r: PairlRecord): string {
  if (r.kind === "marker") {
    if (r.parent !== undefined) return `#msg ${r.name} r=${r.role} parent=${r.parent}`;
    return `#${r.name}`;
  }
  let head: string;
  if (r.kind === "intent") {
    const params = Object.entries(r.kv).map(([k, v]) => `${k}=${fmtValue(v)}`).join(",");
    head = `${r.name}{${params}}`;
  } else if (r.kind === "s") {
    head = r.arg ? `#s ${r.arg}` : "#s";
  } else {
    const pairs = Object.entries(r.kv).map(([k, v]) => `${k}=${fmtValue(v)}`).join(" ");
    head = `#${r.name}${pairs ? " " + pairs : ""}`;
  }
  let trail = "";
  if (r.m) trail += ` @m=${r.m}`;
  if (r.rid) trail += ` @rid=${r.rid}`;
  return head + trail;
}

export function canonicalize(msg: Message, forHash = true): string {
  const lines: string[] = [];
  const present = Object.keys(msg.headers);
  const ordered = HEADER_ORDER.filter((k) => k in msg.headers);
  ordered.push(...present.filter((k) => !HEADER_ORDER.includes(k) && k !== "hash").sort());
  if (!forHash && "hash" in msg.headers) ordered.push("hash");
  for (const k of ordered) lines.push(`@${k} ${msg.headers[k]}`);
  lines.push("");
  for (const r of msg.records) lines.push(serializeRecord(r)); // columnar already expanded
  return lines.join("\n") + "\n";
}

export function computeHash(msg: Message): string {
  return createHash("sha256").update(canonicalize(msg, true), "utf8").digest("hex");
}

export function hashRef(msg: Message): string {
  return `ref:hash:sha256:${computeHash(msg)}`;
}

File: impl/python/pairl/canonical.py
"""Canonicalization and hashing (SPEC §9).

Columnar blocks are expanded to `#type key=value` records before
canonicalization (§9.4a), so a message hashes identically in either form.
"""

from __future__ import annotations

import hashlib

from .core import Message, Record

# Canonical header order (§9.1). Unknown headers are appended, sorted.
_HEADER_ORDER = ["v", "id", "mid", "sid", "ts", "p", "parent", "root", "deps", "budget", "limit"]
_ATOM_OK = set("ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789:._/@+-")


def _needs_quote(v: str) -> bool:
    return v == "" or any(c not in _ATOM_OK for c in v)


def _fmt_value(v: str) -> str:
    if _needs_quote(v):
        return '"' + v.replace('"', '\\"') + '"'
    return v


def serialize_record(r: Record) -> str:
    if r.kind == "marker":
        if r.parent is not None:  # verbose #msg
            return f"#msg {r.name} r={r.role} parent={r.parent}"
        return f"#{r.name}"

    if r.kind == "intent":
        params = ",".join(f"{k}={_fmt_value(v)}" for k, v in r.kv.items())
        head = f"{r.name}{{{params}}}" if r.kv or True else r.name
    elif r.kind == "s":
        head = f"#s {r.arg}" if r.arg else "#s"
    else:
        pairs = " ".join(f"{k}={_fmt_value(v)}" for k, v in r.kv.items())
        head = f"#{r.name} {pairs}".rstrip()

    trail = ""
    if r.m:
        trail += f" @m={r.m}"
    if r.rid:
        trail += f" @rid={r.rid}"
    return head + trail


def canonicalize(msg: Message, *, for_hash: bool = True) -> str:
    """Produce the canonical text. With for_hash, @hash is omitted (§9.6)."""
    lines: list[str] = []
    keys = list(msg.headers.keys())
    ordered = [k for k in _HEADER_ORDER if k in msg.headers]
    ordered += sorted(k for k in keys if k not in _HEADER_ORDER and k != "hash")
    if not for_hash and "hash" in msg.headers:
        ordered.append("hash")
    for k in ordered:
        lines.append(f"@{k} {msg.headers[k]}")

    lines.append("")  # blank line separating header block from body

    # Records already include columnar rows expanded into Record objects (§9.4a),
    # so iterating msg.records yields the canonical, expanded body.
    for r in msg.records:
        lines.append(serialize_record(r))

    return "\n".join(lines) + "\n"


def compute_hash(msg: Message) -> str:
    canon = canonicalize(msg, for_hash=True)
    digest = hashlib.sha256(canon.encode("utf-8")).hexdigest()
    return digest


def hash_ref(msg: Message) -> str:
    return f"ref:hash:sha256:{compute_hash(msg)}"
